internal/c/parts/audio/audio.cpp

```diff
@@ -93,337 +71,286 @@ extern ptrszint dblock;         // Required for Play(). Did not find this declar
+/// @brief A simple FP32 stereo sample frame
+struct SampleFrame {
+    float l;
+    float r;
 };
 
-/// <summary>
-/// This is a light weight queue of type SampleFrameBlockNode and also the guts of SndRaw.
-/// We could have used some std container but I wanted something really lean, simple and transparent.
-/// </summary>
-struct SampleFrameBlockQueue {
-    SampleFrameBlockNode *first;           // First sample frame block
-    SampleFrameBlockNode *last;            // Last sample frame block
-    size_t blockCount;                     // Sample frame block count
-    size_t frameCount;                     // Number of sample frames we have in the queue
-    ma_uint32 sampleRate;                  // The sample rate reported by ma_engine
-    ma_uint32 blockSampleFrames;           // How many sample frames do we need per 'block'. See below
-    float *buffer;                         // This is the ping-pong buffer where the samples block will be 'streamed' to
-    ma_uint32 bufferSampleFrames;          // Size of the ping-pong buffer in *samples frames*
-    bool updateFlag;                       // We will only update the buffer with fresh samples when this flag is not equal to the check condition
-    ma_uint32 bufferUpdatePosition;        // The position (in samples) in the buffer where we should be copying a sample block
-    ma_uint32 sampleFramesPlaying;         // The number of sample frames that was sent for playback
-    ma_uint64 maEngineTime;                // miniaudio engine time use for correct length calculation
-    ma_sound *maSound;                     // Pointer to a ma_sound object that was passed in the constructor
-    ma_engine *maEngine;                   // Pointer to a ma_engine object
-    ma_audio_buffer maBuffer;              // miniaudio buffer object
-    ma_audio_buffer_config maBufferConfig; // miniaudio buffer configuration
-    ma_result maResult;                    // This is the result of the last miniaudio operation (used for trapping errors)
-
-    SampleFrameBlockQueue() = delete;                                   // No default constructor
-    SampleFrameBlockQueue(const SampleFrameBlockQueue &) = delete;      // No default copy constructor
-    SampleFrameBlockQueue &operator=(SampleFrameBlockQueue &) = delete; // No assignment operator
-
-    /// <summary>
-    /// This initializes the queue and calculates the sample frames per block
-    /// </summary>
-    /// <param name="pmaEngine">A pointer to a miniaudio engine object</param>
-    /// <param name="pmaSound">A pointer to a miniaudio sound object</param>
-    SampleFrameBlockQueue(ma_engine *pmaEngine, ma_sound *pmaSound) {
-        first = last = nullptr;
-        blockCount = frameCount = maEngineTime = sampleFramesPlaying = bufferUpdatePosition = 0;
+/// @brief A miniaudiio raw audio stream datasource
+struct RawStream {
+    ma_data_source_base maDataSource;         // miniaudio data source (this must be the first member of our struct)
```

Comment:
I was worried about that too. But since we are not using memset on the struct objects we should be good. I checked the code on the miniaudio side too and it does not do anything funny. I compiled and tested this on VC++ and it worked there as well.

Reply:
Well I was thinking that because there's usually a VTable pointer placed at the beginning of a class/struct in C++ the pointer to `ma_data_source_base` wouldn't be directly convertible to a `RawStream`. But if it wasn't working I would expect at least the `RawStreamOnRead()` callback to explode, so I'm probably missing something, Maybe this class/struct doesn't have a VTable? I don't know the details on when C++ would add one, you don't have any `virtual` functions so maybe that's why.

We have tests that play actual sounds so I think we would notice if it was exploding :D

Reply:
Yes. A vtable will be added when the class has virtual functions.